### ISSUE TITLE
Question about sd3_sde_with_logprob implementation

### ISSUE THREAD

**Issue by thechargedneutron** [yifan123/flow_grpo#32]
Hi,

Thanks for the code. I have a few questions regarding `sd3_sde_with_logprob` implementation.

1. How did you choose 0.7 [here](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_sde_with_logprob.py#L47)? And how do I decide this number for a different model that I plan to use?
2. The `dt` calculated comes to be negative, and that is why you use `-1*dt` when calculating the `sqrt`. In a simple case when you set `0.7` to `→ 0` in point 1 above, we get `prev_sample = sample + dt * model_output` which is the same as the deterministic case, but since your `dt` is negative, should this equation not be different? i.e., should it not be `sample - dt * model_output`?

I am new to this field so may have missed something, thanks in advance.

**Comment by yifan123** [comment]
If the result from deterministic sampling with a=0 is abnormal, it’s likely that the directions of `v` and `dt` are incorrect. You can share a minimal sampling script in the issue, and I’ll take a look.

**Comment by dgr9527** [comment]
When a=0, the result is equally bad. When the function sde_step_with_logprob is not used, the sampling result is normal.

**Comment by dgr9527** [comment]
When a=0, the result is equally bad. When the function sde_step_with_logprob is not used, the sampling result is normal.

def sde_step_with_logprob_flux(
    self: FlowMatchEulerDiscreteScheduler,
    model_output: torch.FloatTensor,
    timestep: Union[float, torch.FloatTensor],
    sample: torch.FloatTensor,
    prev_sample: Optional[torch.FloatTensor] = None,
    generator: Optional[torch.Generator] = None,
    determistic: bool = False
) -> Union[FlowMatchEulerDiscreteSchedulerOutput, Tuple]:
    """
    Predict the sample from the previous timestep by reversing the SDE. This function propagates the flow
    process from the learned model outputs (most often the predicted velocity).

    Args:
        model_output (`torch.FloatTensor`):
            The direct output from learned flow model.
        timestep (`float`):
            The current discrete timestep in the diffusion chain.
        sample (`torch.FloatTensor`):
            A current instance of a sample created by the diffusion process.
        generator (`torch.Generator`, *optional*):
            A random number generator.
    """
    step_index = [self.index_for_timestep(t) for t in timestep]
    prev_step_index = [step+1 for step in step_index]
    sigma = self.sigmas[step_index].view(-1, 1, 1)
    sigma_prev = self.sigmas[prev_step_index].view(-1, 1, 1)
    sigma_max = self.sigmas[1].item()
    dt = sigma_prev - sigma

    std_dev_t = torch.sqrt(sigma / (1 - torch.where(sigma == 1, sigma_max, sigma)))*0  # TODO
    
    # our sde
    prev_sample_mean = sample*(1+std_dev_t**2/(2*sigma)*dt)+model_output*(1+std_dev_t**2*(1-sigma)/(2*sigma))*dt
    
    if prev_sample is not None and generator is not None:
        raise ValueError(
            "Cannot pass both generator and prev_sample. Please make sure that either `generator` or"
            " `prev_sample` stays `None`."
        )

    if prev_sample is None:
        variance_noise = randn_tensor(
            model_output.shape,
            generator=generator,
            device=model_output.device,
            dtype=model_output.dtype,
        )
        prev_sample = prev_sample_mean + std_dev_t * torch.sqrt(-1*dt) * variance_noise

    # No noise is added during evaluation
    if determistic:
        prev_sample = sample + dt * model_output

    log_prob = (
        -((prev_sample.detach() - prev_sample_mean) ** 2) / (2 * ((std_dev_t * torch.sqrt(-1*dt))**2))
        - torch.log(std_dev_t * torch.sqrt(-1*dt))
        - torch.log(torch.sqrt(2 * torch.as_tensor(math.pi)))
    )

    # mean along all but batch dimension
    log_prob = log_prob.mean(dim=tuple(range(1, log_prob.ndim)))

    return prev_sample, log_prob, prev_sample_mean, std_dev_t * torch.sqrt(-1*dt)

**Comment by thechargedneutron** [comment]
I don't have it right now, can you tell the exact model you are using and I will take a look?

**Comment by dgr9527** [comment]
> > I don't have it right now, can you tell the exact model you are using and I will take a look?
> 
> flux-dev The code is in the previous reply.

@thechargedneutron

**Comment by leckiepeng** [comment]
@yifan123 So is this a typo? should be |delta_t| in the sqrt.

<img width="880" height="178" alt="Image" src="https://github.com/user-attachments/assets/f2b7a751-6c95-4c52-a64c-1734df9631bc" />
